Feat/FE-BE  Replace user id placeholder to actual user id to add and remove favrotite movie

Kositthai opened the pull request:
# Description

- Replace user id placeholder to actual user id
- Add and remove favorite working with actual user id
- Create getUserInformation endpoint

# Jira task link
https://hackyourfuture-dk.atlassian.net/jira/software/projects/LANDING/boards/6?selectedIssue=LANDING-80

# Screenshots (if applicable)

shobanamg commented:
Hi @Kositthai, Can you adjust this PR, so that the userId is set in the user context so that the userId is available throughout the application?  FYI @danJecu, @shpomp

magdazelena commented on packages/client/src/components/MovieCard/useFavorites.js:
```diff
@@ -6,22 +6,40 @@ import { useUserContext } from '../../context/UserContext';
   const navigate = useNavigate();
 
+  useEffect(() => {
+    const getUserIdByUid = async () => {
+      if (user && user.uid) {
+        try {
+          const response = await fetch(`${apiURL()}/users/${user.uid}`);
+          const data = await response.json();
+          setUserId(data[0].id);
+        } catch (error) {
+          return error.name === 'Something wrong in the operation';
+        }
+      }
+    };
+
+    getUserIdByUid();
+  }, [user]);
```
Hm, it's interesting that userId is not in user context and that there are two separate ids? If I understand correctly :)

Perhaps it would be nice to streamline this within userContext. Otherwise this snippet is obv correct :)

magdazelena commented on packages/client/src/components/MovieCard/useFavorites.js:
```diff
@@ -6,22 +6,40 @@ import { useUserContext } from '../../context/UserContext';
-        setFavorites(data);
-      } catch (error) {
-        return error.name === 'this is an aborted error';
+      if (user && user.uid) {
```
why are you checking for those here?

magdazelena commented on packages/server/api/controllers/users.controller.js:
```diff
@@ -33,6 +33,11 @@ const createUser = async (body) => {
 };
 
+const getUserInformation = (uID) => {
+  return knex('users').select('id').where({ uid: uID });
```
oh, I know why you have two different ids now :)

Usually it is better to not expose database uuid's if there is another ID to do with. Or just stick to one id per db entry.

magdazelena commented on packages/server/api/routes/users.router.js:
```diff
@@ -13,4 +13,22 @@ router.post('/', async (req, res, next) => {
+    .getUserInformation(uid)
+    .then((result) => {
+      if (result.length === 0) {
+        res.status(404).send('Invalid user  id');
```
picking on details here, but with 404 it should be more "couldn't find user with this id".

Ivalid means some syntax/parsing error of the string itself.

magdazelena reviewed (changes requested):
In general this is good code, but I would encourage to streamline the id situation. This might be also for team discussion how and why :)

Kositthai commented:
> In general this is good code, but I would encourage to streamline the id situation. This might be also for team discussion how and why :)

Thank you so much for your review!
I have edited my code as your suggestions and make a few changed 

- I have moved my getUserIdByUid to userContext now, so our teams can just import **userId** in thier file and use it 

- I changed an error msg in users router like your suggestion 

- I have moved some checking invalidation user Id format from favorite router to favorite controller based on other PR that I get recommended to change this too 

- Changed the path if the user is not log in when click on heart icon, then redirect to signin page 

**why are you checking for those here? if(user && user.uid)** 
Because I want to make a fetch only when the user is login so I can use user.uid to fetch user_id from DB.

Kositthai commented:
> Hi @Kositthai, Can you adjust this PR, so that the userId is set in the user context and it is available throughout the application? FYI @danJecu, @shpomp

Thank you so much for bring this up! I totally didnt think about it at all
If this PR is merged without other changes, you can access to userId for your page. _You can also look at my useFavorite function as an example_

magdazelena reviewed (approved):
Good work! I would still encourage to clear up two-separate-ids situation, but it perhaps is a subject another PR/nice to have :)

Kositthai commented:
> Good work! I would still encourage to clear up two-separate-ids situation, but it perhaps is a subject another PR/nice to have :)

Thank you so much! but I still dont understand what do you mean clear up two-separate-ids?

shobanamg commented:
> > Hi @Kositthai, Can you adjust this PR, so that the userId is set in the user context and it is available throughout the application? FYI @danJecu, @shpomp
> 
> @shobanamg Thank you so much for bring this up! I totally didnt think about it at all If this PR is merged without other changes, you can access to userId for your page. _You can also look at my useFavorite function as an example_

Actually, we don't need useFavorite hook at all. I see that you guys are fetching favorites separately. In our case movie's end point by default returns the is_favorite flag if you pass the user_id to it. You can have a look at how we do it. But anyway Thanks.
